cmd/snapd-apparmor/main.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/snapcore/snapd/logger"
+	apparmor_sandbox "github.com/snapcore/snapd/sandbox/apparmor"
+	"github.com/snapcore/snapd/snapdtool"
+)
+
+// Checks to see if the current container is capable of having internal AppArmor
+// profiles that should be loaded.
+//
+// The only known container environments capable of supporting internal policy
+// are LXD and LXC environment.
+//
+// Returns true if the container environment is capable of having its own internal
+// policy and false otherwise.
+//
+// IMPORTANT: This function will return true in the case of a non-LXD/non-LXC
+// system container technology being nested inside of a LXD/LXC container that
+// utilized an AppArmor namespace and profile stacking. The reason true will be
+// returned is because .ns_stacked will be "yes" and .ns_name will still match
+// "lx[dc]-*" since the nested system container technology will not have set up
+// a new AppArmor profile namespace. This will result in the nested system
+// container's boot process to experience failed policy loads but the boot
+// process should continue without any loss of functionality. This is an
+// unsupported configuration that cannot be properly handled by this function.
+//
+func isContainerWithInternalPolicy() bool {
+	var appArmorSecurityFSPath = filepath.Join(dirs.GlobalRootDir, "/sys/kernel/security/apparmor")
+	var nsStackedPath = filepath.Join(appArmorSecurityFSPath, ".ns_stacked")
+	var nsNamePath = filepath.Join(appArmorSecurityFSPath, ".ns_name")
+
+	for _, path := range []string{nsStackedPath, nsNamePath} {
+		if _, err := os.Stat(path); errors.Is(err, os.ErrNotExist) {
+			return false
+		}
+	}
+
+	contents, err := ioutil.ReadFile(nsStackedPath)
+	if err != nil {
+		return false
+	}
+
+	if strings.TrimSpace(string(contents)) != "yes" {
+		return false
+	}
+
+	contents, err = ioutil.ReadFile(nsNamePath)
+	if err != nil {
+		return false
+	}
+
+	// LXD and LXC set up AppArmor namespaces starting with "lxd-" and
+	// "lxc-", respectively. Return false for all other namespace
+	// identifiers.
+	name := strings.TrimSpace(string(contents))
+	if !strings.HasPrefix(name, "lxd-") && !strings.HasPrefix(name, "lxc-") {
+		return false
+	}
+	return true
+}
+
+func loadAppArmorProfiles() error {
+	candidates, err := filepath.Glob(dirs.SnapAppArmorDir + "/*")
+	if err != nil {
+		return err
+	}
+
+	profiles := make([]string, 0, len(candidates))
+	for _, profile := range candidates {
+		// Filter out profiles with names ending with ~, those are
+		// temporary files created by snapd.
+		if strings.HasSuffix(profile, "~") {
+			continue
+		}
+		profiles = append(profiles, profile)
+	}
+	if len(profiles) == 0 {
+		logger.Noticef("No profiles to load")
+		return nil
+	}
+	logger.Noticef("Loading profiles %v", profiles)
+	err = apparmor_interface.LoadProfiles(profiles, apparmor_sandbox.SystemCacheDir, 0)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func isContainer() error {
```

Comment:
minor: it would be more logical for this function to return a boolean

Reply:
Ok, so I assume then we would want to have `isContainer()` swallow any error and return `false` in that case?

Reply:
Done in https://github.com/snapcore/snapd/pull/11129/commits/9b4f081e8df59aa6c05e4b7be7ace74491829e99 - let me know if this is what you had in mind.

Reply:
Yes, thanks!